src/main/java/jenkins/bouncycastle/api/PEMManager.java

```diff
@@ -0,0 +1,394 @@
+     * 
+     * @return public key fingerprint in hex format "ab:cd:ef:...:12", null if the public key could not be obtained from
+     * the current PEM data.
+     * @throws IOException if a problem exists creating the fingerprint
+     */
+    @Nullable
+    public String getPublicKeyFingerprint() throws IOException {
+        PublicKey key = getPublicKey();
+        if (key == null) {
+            return null;
+        }
+        return hexEncode(getKeyDigestMD5(key));
+    }
+
+    /**
+     * Generates an SHA1 digest from a Key object
+     * 
+     * @param k the key to generate the digest from
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
+     */
+    @Nonnull
+    public static byte[] getKeyDigestSHA1(@Nonnull Key k) throws IOException {
+        return getKeyDigest(k, "SHA1");
+    }
+
+    /**
+     * Generates an MD5 digest from a Key object
+     * 
+     * @param k the key to generate the digest from
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
+     */
+    @Nonnull
+    public static byte[] getKeyDigestMD5(@Nonnull Key k) throws IOException {
+        return getKeyDigest(k, "MD5");
+    }
+
+    /**
+     * Generates an digest from a Key object in the specified digest format. The supported digest formats will depend on
+     * the JVM API.
+     * 
+     * @param k key to generate the digest from
+     * @param dg digest format
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
+     */
+    @Nonnull
+    public static byte[] getKeyDigest(@Nonnull Key k, @Nonnull String dg) throws IOException {
+        try {
+            MessageDigest md5 = MessageDigest.getInstance(dg);
+
+            DigestInputStream in = new DigestInputStream(new ByteArrayInputStream(k.getEncoded()), md5);
+            try {
+                while (in.read(new byte[128]) > 0)
+                    ; // simply discard the input
+            } finally {
+                in.close();
+            }
+            return md5.digest();
+        } catch (NoSuchAlgorithmException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    /**
+     * Encode {@link byte[]} in hex formated string "ab:cd:ef:...:12"
+     * 
+     * @param data to be encoded
+     * @return hex formated string "ab:cd:ef:...:12"
+     */
+    public static String hexEncode(byte[] data) {
+        char[] hex = Hex.encodeHex(data);
+        StringBuilder buf = new StringBuilder();
+        for (int i = 0; i < hex.length; i += 2) {
+            if (buf.length() > 0)
+                buf.append(':');
+            buf.append(hex, i, 2);
+        }
+        return buf.toString();
+    }
+
+    /**
+     * Encodes a {@link byte[]} in base 64 string
+     * 
+     * @param data to be encoded
+     * @return base 64 formatted string
+     */
+    public static String encodeBase64(byte[] data) {
```

Comment:
nullness annotations required here. also do we need to expose these as public methods?